### PR TITLE
bpo-41762: Fix usage of productionlist markup in the doc

### PR DESCRIPTION
**Pull request by vstinner**
Use an unique identifier for the different grammars documented using
the Sphinx productionlist markup.

productionlist markups of the same grammar, like "expressions" or
"compound statements", use the same identifier.

https://bugs.python.org/issue41762

### CODE REVIEW

**Comment by vstinner** [comment]
I checked manually the rendered HTML documentation: reference/expressions.html still contains clickable links between the different grammar items. For example, I can click on "expression" from lambda_expr and I'm pointed to expression definition of the "6.13. Conditional expressions" section. I also checked library/string.html: links are kept.

cc @JulienPalard @deronnax @isidentical @pablogsal @lysnikolaou

**Review by pablogsal — APPROVED** [review]

File: Doc/reference/toplevel_components.rst
Comment (vstinner) [inline Doc/reference/toplevel_components.rst:89]: This change breaks the link to stmt_list.

Should we use the same identifier here, in expressions, simple statements and compound statements?

Comment (pablogsal) [reply Doc/reference/toplevel_components.rst:89]: Yeah, I think that is a good idea (to preserve those links). Is there any other missing links between sections?

Comment (pablogsal) [reply Doc/reference/toplevel_components.rst:89]: Or maybe use the least amount of namespaces possible?

Comment (vstinner) [reply Doc/reference/toplevel_components.rst:89]: I didn't check all modified pages.

**Review by corona10 — APPROVED** [review]
@vstinner @pablogsal 

I can check my PR works well after this PR is merged :)
https://github.com/python/cpython/pull/22198

**Comment by vstinner** [comment]
`Doc/reference/*.rst` must all live in the same productionlist domain (same grammar):
* top_level => compound_stmts => simple_stmts => expressions => lexical_analysis

Examples:

* https://docs.python.org/dev/reference/toplevel_components.html#interactive-input : compound_stmt of interactive_input links to https://docs.python.org/dev/reference/compound_stmts.html#grammar-token-compound-stmt
* https://docs.python.org/dev/reference/compound_stmts.html#grammar-token-stmt-list : simple_stmt  of stmt_list links to: https://docs.python.org/dev/reference/simple_stmts.html#grammar-token-simple-stmt
* https://docs.python.org/dev/reference/simple_stmts.html#grammar-token-expression-stmt : starred_expression of expression_stmt links to https://docs.python.org/dev/reference/expressions.html#grammar-token-starred-expression
* https://docs.python.org/dev/reference/expressions.html#grammar-token-atom : identifier of atom links to: https://docs.python.org/dev/reference/lexical_analysis.html#grammar-token-identifier

**Comment by vstinner** [comment]
Ok, I pushed a 3rd commit.

I checked links between reference/ documents and they all have been repaired by my 3rd commit.

I'm going to merge this PR once the CI pass.

**Pull request merged**
